File: Python/004-largest_palindrome_product.py
# -*- coding: utf-8 -*-
"""
A palindromic number reads the same both ways.
The largest palindrome made from the product of two 2-digit numbers is

    9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.
"""

def is_palindrome(string):
    """ Brute force algorithm """
    decide, i = 1, 0
    while (i<=int(len(string)/2) and decide==1):
        if string[i] != string[-(i+1)]:
            decide=0
        i+=1
    return decide

def main(digits):
    aux=0
    for i in range(10^(digits-1)+1, 10**digits):
        for j in range(10^(digits-1)+1, 10**digits):
            if (is_palindrome(str(j*i)) and i*j>aux):
                aux=i*j
    return aux

if __name__ == '__main__':
    print(main(3))

File: C/004-largest_palindrome_product.c
/*
 *
 * A palindromic number reads the same both ways.
 * The largest palindrome made from the product of two 2-digit numbers is
 *
 *    9009 = 91 × 99.
 *
 * Find the largest palindrome made from the product of two 3-digit numbers.
 *
 */

#include <stdio.h>
#include <string.h>
#include "itoa.h"

int is_palindrome(char* string) {
    int i, palindrome = 1;

    for (i=0; i<=(int)(strlen(string))/2; i++)
        if (string[i] != string[(int)(strlen(string))-i-1])
            palindrome = 0;

    return palindrome;
}

int main(void)
{
    char c[8];
    int i, j, solution=0;

    for (i=100; i<=1000; i++)
        for (j=100; j<=1000; j++) {
            itoa(i*j, c);
            if (is_palindrome(c) && i*j>solution)
                solution=i*j;
    }

    printf("Solution is %i\n", solution);
    return 0;
}

